### ISSUE TITLE
evhttp_connection_free causes segmentfault because of double free

### ISSUE THREAD

**Issue by LinkShen** [libevent/libevent#997]
- Environment
os: debian8
libevent: 2.1.8

- Problem
Our project has a simply http server implemented with libevent http library. Few days ago, one of our http server processes crashed at the line evhttp_connection_free. The case is a user sent a GET request to our http server, before we can send a reply, the user client closed the connection. For some reason, our http server has our own timeout(not libevent http timeout), when this request timeout, we did some clean up jobs include free evhttp_connection and evhttp_request.
We tried to reproduce the crash afterward, and we found another odd issue, if we send a request with bodylen > 0 and close the connection before reply, then it won't cause double free when timeout. However when we send a request without body, then we can reproduce the crash.
We found some evidence in http.c of libevent:

1. request without body
```
static void
evhttp_error_cb(struct bufferevent *bufev, short what, void *arg)
{
    ...
    else if (what & (BEV_EVENT_EOF|BEV_EVENT_ERROR)) {
        ...
        evhttp_connection_fail_(evcon, EVREQ_HTTP_EOF);
    }
    ...
}

void
evhttp_connection_fail_(struct evhttp_connection *evcon,
    enum evhttp_request_error error)
{
	...
	if (evcon->flags & EVHTTP_CON_INCOMING) {
		/*
		 * for incoming requests, there are two different
		 * failure cases.  it's either a network level error
		 * or an http layer error. for problems on the network
		 * layer like timeouts we just drop the connections.
		 * For HTTP problems, we might have to send back a
		 * reply before the connection can be freed.
		 */
		if (evhttp_connection_incoming_fail(req, error) == -1)
			evhttp_connection_free(evcon);
		return;
	}
        ...
}
```
According to the code libevent free the connection when detect client close connection. So it cause double free(libevent free first, then our code free the second time).

2. request with body
```
static void
evhttp_read_body(struct evhttp_connection *evcon, struct evhttp_request *req)
{
        ...
	if (!req->ntoread) {
		bufferevent_disable(evcon->bufev, EV_READ);
		/* Completed content length */
		evhttp_connection_done(evcon);
		return;
	}
}
```
According to the code, when request has body, libevent will disable EV_READ(this not occur when no body), so when the client close connection the code won't enter evhttp_error_cb, then libevent won't free connection, then there is no double free. (Is this mean if we don't use evhttp_connection_free, there is a memory leak under this case?)

- Question
Do we really need to cleanup evhttp_connection or evhttp_request by ourself? If not, is libevent handle all condition that need free these data struct? And why libevent provide evhttp_connection_free and evhttp_request_free for us?
Thx for your attention and look forward to your reply. ORZ

**Comment by azat** [comment]
>According to the code, when request has body, libevent will disable EV_READ(this not occur when no body), so when the client close connection the code won't enter evhttp_error_cb, then libevent won't free connection, then there is no double free. (Is this mean if we don't use evhttp_connection_free, there is a memory leak under this case?)

Yes, but it will be enabled once the response will be send, and it will got EOF and destroy the connection

>Do we really need to cleanup evhttp_connection or evhttp_request by ourself? 

In general - no

>  If not, is libevent handle all condition that need free these data struct? 

library structures - yes, yours - not

>And why libevent provide evhttp_connection_free and evhttp_request_free for us?

- `evhttp_connection_free` should be used manually for outgoing connections (in other words when you have client)
- `evhttp_requst_free` only when you change the ownership of the request using `evhttp_request_own` (or if did not call `evhttp_send_reply*`)

P.S. documentation does not says anything about this, should be updated.

**Comment by LinkShen** [comment]
Thanks for your valuable advises. However I still have a question on this issue. 

> Yes, but it will be enabled once the response will be send, and it will got EOF and destroy the connection

Our project cannot guarantee to send a reply under all conditions, is that mean if we don't free the outcoming connection there still have memory leak risk?

**Comment by ploxiln** [comment]
If your project can *try* to send a reply under all conditions, then evhttp should take care of its side of noticing the client is already gone and freeing what it is responsible for.

**Comment by LinkShen** [comment]
Got it. According to all these advise, the best practise for dealing incoming requests is to leave evhttp_connection and evhttp_request to libevent, and make sure to reply all requests even if the connection is lost.
Thx all and I'll close this issue.

**Comment by tonso** [comment]
@azat So if `evhttp_request_own` is called in a handler for incoming request, then I should either call `evhttp_send_reply*` (which will free the request regardless of its ownership), or `evhttp_request_free` explicitly? And what if the connection is terminated, do the user-owned requests also get freed automatically?

**Comment by azat** [comment]
>@azat So if evhttp_request_own is called in a handler for incoming request, then I should either call evhttp_send_reply* (which will free the request regardless of its ownership), or evhttp_request_free explicitly? And what if the connection is terminated, do the user-owned requests also get freed automatically?

`evhttp_request_own` makes sense only for outgoing requests (if you have client)
for incoming requests (if you have server) the request will be freed regardless `evhttp_request_own` after the request was sent

and for incoming requests you should always call evhttp_send_reply (since it do other things except for freeing request)

**Comment by tonso** [comment]
@azat Ok, then is it possible at all to pass an incoming request to another thread for handling (and pass it back later to the main thread, which will call `evhttp_send_reply`)? If there's no way to acquire ownership for incoming requests, and let's say the connection is aborted (and the associated requests are freed immediately), then another thread may be accessing the request while it's being freed, which will cause memory corruption.

**Comment by azat** [comment]
>then is it possible at all to pass an incoming request to another thread for handling (and pass it back later to the main thread, which will call evhttp_send_reply)?

This is highly not recommended with libevent.

>If there's no way to acquire ownership for incoming requests, and let's say the connection is aborted (and the associated requests are freed immediately)

But this should not happen, since the close detection should be disabled (so server will not know that the client goes away until evhttp_send_reply will be called)

**Comment by tonso** [comment]
> This is highly not recommended with libevent.

But if I extract (copy) all the information from a request and pass it to another thread, and once it's processed I notify the main thread, that it can send the result, then this is safe I suppose?

> But this should not happen, since the close detection should be disabled (so server will not know that the client goes away until evhttp_send_reply will be called)

So as far as I understand if the connection gets closed by client after receiving the request but before the response is sent, the requests won't get freed, correct? Is the request get freed on some other abnormal event (connection timeout expires, anything else?)?

**Comment by azat** [comment]
>But if I extract (copy) all the necessary information from a request and pass it to another thread, and once it's processed I notify the main thread that it can send the result, then this is safe I suppose?

If the main thread still in the callback - 100% safe yes

>So the request isn't freed until the the response is sent, but what about abnormal terminations (connection timeout expiration, etc.)? 

Also should not be freed (but this should be verified)

>Is there any hook, which can tell that the requests were (or will be) freed (probably evhttp_connection_set_closecb)?

Right now there is no hook for this.

And note that such design is not good for libevent's http layer, do you really need this? Maybe worth starting multiple http instances w/ SO_REUSEPORT and do not use inter-thread communication from the callbacks?

**Comment by tonso** [comment]
@azat 
> do you really need this? Maybe worth starting multiple http instances w/ SO_REUSEPORT and do not use inter-thread communication from the callbacks?

I'm ok with anything that won't block the whole server with a single request. Do you mean creating multiple event bases with a dedicated http instance per base?

**Comment by azat** [comment]
>I'm ok with anything that won't block the whole server with a single request. Do you mean creating multiple event bases with a dedicated http instance per base?

Yes, so basically if you have heavy callbacks then, you need to use thread pool with:
- separate event_base for each thread
- separate http server instance for each thread

But when you are using SO_REUSEPORT, you are depending from the kernel on how it will schedule the sockets, if you need something more enhanced (and maybe even non-static thread pool), then you can add a scheduler in front, that will call accept and pass the fd to non-busy thread or create more threads

>But if I postpone handling of the incoming request (using a timer for example), then it can be freed if an error occurs (via evhttp_error_cb -> evhttp_connection_fail_), correct?

It should not, remember events should be disabled, so it cannot detect the close (but again this need to be verified)
